src/win32/console.cpp

```diff
@@ -20,63 +20,645 @@
+    case HistorySearchForward: searchHistory(true);  break;
+    case HistorySearchBackward: searchHistory(false);  break;
+    case ClearLine: searchingHistory ? deleteHistorySearchChars(0) : deleteCharRange(0, bufSize);  break;
+    case DeleteCharLeft: searchingHistory ? deleteHistorySearchChars(1) : deleteCharRange(pos - 1, pos);  break;
+    case DeleteCharRight: deleteCharRange(pos, pos + 1); break;
+    case DeleteWordLeft: deleteCharRange(detectWordBoundary(pos, false), pos);  break;
+    case DeleteWordRight: deleteCharRange(pos, detectWordBoundary(pos, true));  break;
+    case AutoCompleteForwards: autoComplete(true, consoleWidth); break;
+    case AutoCompleteBackwards: autoComplete(false, consoleWidth); break;
+    }
+}
+
+bool ConsoleModel::checkForCompletedInputLine(std::wstring& ws)
+{
+    auto newlinePos = std::find(buffer.begin(), buffer.end(), 13);
+    if (newlinePos != buffer.end())
+    {
+        ws.assign(buffer.begin(), newlinePos);
+        buffer.erase(buffer.begin(), newlinePos + 1);
+        insertPos = 0;
+        newlinesBuffered -= 1;
+        bool sameAsLastCommand = !inputHistory.empty() && inputHistory[0] == ws;
+        bool sameAsChosenHistory = !inputHistory.empty() &&
+            inputHistoryIndex >= 0 && inputHistoryIndex < inputHistory.size() &&
+            inputHistory[inputHistoryIndex] == ws;
+        if (echoOn && !sameAsLastCommand && !ws.empty())
+        {
+            if (inputHistory.size() + 1 > MaxHistoryEntries)
+            {
+                inputHistory.pop_back();
+            }
+            inputHistory.push_front(ws);
+            inputHistoryIndex = sameAsChosenHistory ? inputHistoryIndex + 1 : -1;
+        }
+        enteredHistory = false;
+        return true;
+    }
+    return false;
+}
+
+
 WinConsole::WinConsole()
 {
-    // FIXME: configure for UTF-8 I/O
+    hInput = GetStdHandle(STD_INPUT_HANDLE);
+    hOutput = GetStdHandle(STD_OUTPUT_HANDLE);
+
+    DWORD dwMode;
+    GetConsoleMode(hInput, &dwMode);
+    SetConsoleMode(hInput, dwMode & ~(ENABLE_MOUSE_INPUT));
+    FlushConsoleInputBuffer(hInput);
 }
 
 WinConsole::~WinConsole()
 {
     // restore startup config
 }
 
-void WinConsole::readpwchar(char* pw_buf, int pw_buf_size, int* pw_buf_pos, char** line)
+void WinConsole::setShellConsole()
 {
-    char c;
-    DWORD cread;
+    // Call this if your console app is taking live input, with the user editing commands on screen, similar to cmd or powershell
+
+    // use cases covered
+    // utf8 output with std::cout (since we already use cout so much and it's compatible with other platforms)
+    // unicode input with windows ReadConsoleInput api
+    // drag and drop filenames from explorer to the console window
+    // upload and download unicode/utf-8 filenames to/from Mega
+    // input a unicode/utf8 password without displaying anything
+    // normal cmd window type editing (but no autocomplete, as yet - we could add it now though)
```

Comment:
autocomplete is working too now, isn't it?

Reply:
yes, thanks for spotting that.  Comments go out of date so fast :-).  I've updated them.